File: dsp/lecture_LP.m
% create time-based signal
% define sampling frequency
% frequency in Hertz
Fsamp = 125;

%define the period
T = 1/Fsamp;

%set the time base
t = 0:T:1;

% define signal frequency
Fs = 50;

% define signal
signal = cos(2*pi*Fs*t);

figure;
plot(t, signal);
%xlim([0 250]);
ylim([-1.1 1.15]);
title("signal");

% define noise frequency
Fn = 5;

% define noise
noise = 0.2*cos(2*pi*Fn*t);

figure;
plot(t, noise);
%xlim([0 500]);
ylim([-1.1 1.15]);
title("noise");

waveform = signal + noise;
figure;
plot(t, waveform);
title("waveform");
%xlim([0 50]);
ylim([-1.25 1.25]);



% cutoff frequency is defined with respect to normalized nyquist frequency
% if sampling is Fs then normalized is 1
% then normalized nyquist is 0.5
% this means that highest frequency in signal cannot exceed 0.5*Fs
% 
% set the cutoff frequency
Fc = 12.5;
freq_cutoff = Fc/Fsamp;

% set the limits of the sinc even though it is infinite
sLimit = 25;
n = -sLimit:sLimit;

%{
y = sinc(n*freq_cutoff);
figure;
stem(n,  y);
ylim([-0.4 1.1]);
%}

% define the impluse response fo the filter -- for low pass
% this is a sinc filter in the time domain
h = theSinc(n, 2*freq_cutoff);
figure;
stem(n,  h);
ylim([-0.4 1.1]);
title("theSinc");

% define the blackman window to truncate the impulse response
% window length
N = sLimit*2;
bn = 0:N;
blackmanWindow = 0.42 - 0.5*cos(2*pi*bn/(N-1)) + 0.08*cos(4*pi*bn/(N-1));
figure;
stem(bn, blackmanWindow);
ylim([-0.4 1.1]);
title("blackmanWindow");

hBlackman = h.*blackmanWindow;
figure;
stem(n, hBlackman);
ylim([-0.4 1.1]);
title("hBlackman");

% create normalized impulse response
hNorm = hBlackman/theSum(hBlackman);
figure;
stem(bn,  hNorm);
title("hNorm");

HN = fft(hNorm);
HNs = fftshift(HN);
figure;
plot(abs(HN));
%xlim([0 25]);
title("Norm freq spectrum");
figure;
semilogy(abs(HN));
%xlim([0 500]);
title("Norm freq spectrum dB");

% generate output by convolving waveform and filter

output = dspConv(waveform, hNorm);
%ouput = conv(waveform, hNorm);
figure;
to = 1: length(output);
plot(to, output);
title("output");
%xlim([0 500]);
ylim([-1.25 1.25]);

% ==========
% MARK: dsp functions
function s = theSinc(n, freqCutoff)
for i = 1:length(n);
        sx = freqCutoff*n(i);
        if n(i) == 0
            s(i) = 1;
        else
            s(i) = sin(pi*sx) / (pi*sx);
        end
end
end

function s = theSum(f)
    n = length(f);
    s = 0;
        for i = 1:n
            s = s + f(i);
        end
end

% convolution of filter and input (x), filter (h) -> output (out)
% two signals result in new vector from convolution operation
function out = dspConv(x, h)
    % x = signal, h = filter
    % Transform the vectors x and h in new vectors with the same length
    X = [x,zeros(1,length(h))];
    H = [h,zeros(1,length(x))];
    % the length of output signal of a convolution operation
    % is the sum of the two input signal minus 1
    for i = 1:length(h)+length(x)-1;
        % Create ouput signal out
        out(i) = 0;
        % FOR Loop to walk through the vector F ang G
        for j = 1:length(x);
            if( i- j + 1 > 0)
            out(i) = out(i) + X(j) * H(i - j + 1);
            else
            end
        end
    end
end
    
%{
for d = 1:10;
    x = 0:pi/d:2*pi;
    y = cos(x);
    figure;
    stem(x, y);
    hold on;
    plot(x, y);
    hold off;
end
%}

File: dsp/lecture_LP.swift
func theSinc(n: [Double], freqCutoff: Double) -> [Double] {
  var s = [Double](repeating: 0, count: n.count)
  for i in 0...(n.count-1) {
    let sx = freqCutoff*n[i]
    if n[i] == 0 {
      s[i] = 1
    } else {
      s[i] = _sin(Double.pi * sx) / (Double.pi * sx);
    }
  }
  return s
}

func dspConv(x: [Double], h: [Double]) -> [Double] {
  // x = signal, h = filter
  // Transform the vectors x and h in new vectors with the same length
  var X = x + [Double](repeating: 0, count: h.count)
  var H = h + [Double](repeating: 0, count: x.count)

  // the length of output signal of a convolution operation
  // is the sum of the two input signal minus 1
  let len = (h.count + x.count - 1)
  var out = [Double](repeating: 0, count: len)
  for i in 0...(h.count + x.count - 2) {
    for j in 0 ... (x.count - 1) {
      if (i - j + 1) > 0 {
      out[i] = out[i] + X[j] * H[i - j + 1]
      }
    }
  }
  return out
}
//frequency in Hertz
let Fsamp = 125.0

//define the period
let T = 1.0/Fsamp

//set the time base
var t = Array(stride(from:0.0, to: 1.0, by: T))

//define signal frequency
let Fs = 50.0

//define the signal
let signal = t.map{_cos(2*Double.pi*Fs*$0)}

//define noise frequency
let Fn = 5.0;

//define noise
let noise = t.map{_cos(2*Double.pi*Fn*$0)}

let waveform = zip(noise, signal).map{$0 + $1}

//set the cutoff frequency
let Fc = 12.5
let freq_cutoff = Fc/Fsamp;

//set the limits of the sinc even though it is infinite
let sLimit = 25.0;
var n = Array(stride(from: -1.0*sLimit, to: sLimit, by: 1))

//define the impulse resonse for the filter -- for low pass
//this is a sinc filter in the time domain
let h = theSinc(n: n, freqCutoff: 2*freq_cutoff)

//define the blackman window to truncate the impulse response
//window length
let N = sLimit*2.0;
let bn = Array(stride(from: 0.0, to: N, by: 1.0))

let blackmanWindow = bn.map{0.42 - 0.5*_cos(2*Double.pi*$0/(N-1)) +
                                  0.08*_cos(4*Double.pi*$0/(N-1))}
let hBlackMan = zip(h, blackmanWindow).map{$0 * $1}
let hBlackManSum = hBlackMan.reduce(0.0, +)
let hNorm = hBlackMan.map{$0 / hBlackManSum}

var result:[Double] = dspConv(x: waveform, h: hNorm)
for elem in result {
  print(elem, terminator: "\n")
}
